File: backend/app/schemas/progress.py
"""
OneStop AI - Progress Tracking Schemas
Pydantic models for the Progress & Analytics module.
"""

from pydantic import BaseModel
from datetime import datetime
from typing import List, Optional


class SkillProgress(BaseModel):
    """A single domain skill score from the latest assessment."""
    domain: str
    score: int
    max_score: int = 100


class RoadmapHistoryItem(BaseModel):
    """Summary of a user's roadmap (active or past)."""
    id: str
    target_role: str
    progress_percentage: int
    milestones_total: int
    milestones_completed: int
    is_active: bool
    created_at: datetime

    class Config:
        from_attributes = True


class AssessmentHistoryItem(BaseModel):
    """A single entry in the user's assessment timeline."""
    id: str
    personality_type: Optional[str] = None
    category_scores: dict
    created_at: datetime

    class Config:
        from_attributes = True


class LearningStats(BaseModel):
    """Aggregate learning statistics across the platform."""
    assessments_taken: int = 0
    roadmaps_created: int = 0
    milestones_completed: int = 0
    milestones_total: int = 0
    courses_recommended: int = 0
    resume_reviews: int = 0
    chat_messages_sent: int = 0


class AchievementResponse(BaseModel):
    """An earned achievement badge."""
    badge_key: str
    badge_name: str
    badge_description: str
    badge_icon: str
    earned_at: datetime

    class Config:
        from_attributes = True


class ProgressOverviewResponse(BaseModel):
    """Complete progress overview combining all tracking dimensions."""
    skill_progress: List[SkillProgress] = []
    roadmap_history: List[RoadmapHistoryItem] = []
    assessment_history: List[AssessmentHistoryItem] = []
    learning_stats: LearningStats
    achievements: List[AchievementResponse] = []

File: frontend/src/types/progress.ts
// ========================
// OneStop AI - Progress Tracking Types
// ========================

export interface SkillProgress {
  domain: string;
  score: number;
  max_score: number;
}

export interface RoadmapHistoryItem {
  id: string;
  target_role: string;
  progress_percentage: number;
  milestones_total: number;
  milestones_completed: number;
  is_active: boolean;
  created_at: string;
}

export interface AssessmentHistoryItem {
  id: string;
  personality_type: string | null;
  category_scores: Record<string, number>;
  created_at: string;
}

export interface LearningStats {
  assessments_taken: number;
  roadmaps_created: number;
  milestones_completed: number;
  milestones_total: number;
  courses_recommended: number;
  resume_reviews: number;
  chat_messages_sent: number;
}

export interface Achievement {
  badge_key: string;
  badge_name: string;
  badge_description: string;
  badge_icon: string;
  earned_at: string;
}

export interface ProgressOverview {
  skill_progress: SkillProgress[];
  roadmap_history: RoadmapHistoryItem[];
  assessment_history: AssessmentHistoryItem[];
  learning_stats: LearningStats;
  achievements: Achievement[];
}
